packages/taquito/src/contract/interface.ts

```diff
@@ -158,7 +158,7 @@ export interface ContractProvider extends StorageProvider {
    *
    * @param OriginationOperation Originate operation parameter
    */
-  originate(contract: OriginateParams): Promise<OriginationOperation>;
+  originate<TContract extends DefaultContractType = DefaultContractType>(contract: OriginateParams<ContractStorageType<TContract>>): Promise<OriginationOperation>;
```

Comment:
I think the same should be added here for origination with batch and wallet:
- https://github.com/ecadlabs/taquito/blob/master/packages/taquito/src/batch/rpc-batch-provider.ts#L115
- https://github.com/ecadlabs/taquito/blob/master/packages/taquito/src/wallet/wallet.ts#L80
- https://github.com/ecadlabs/taquito/blob/master/packages/taquito/src/wallet/wallet.ts#L183

Reply:
batch operations cannot be typed because the method accepts an array of operations:

There is no way to require a specific origination operation type at a specific index of the array (as a typed array). Each item of the array must be any of the possible types.

However, a utility method could be provided for this - see the `example-usages.ts` in the related PR.

Reply:
Oh, I see how .withOrigination works --- looking at that now

Reply:
Add wallet `originate()` and `withOrigination()` types for contract and wallet `batch()`